ethereum/contracts/IncentivizedInboundChannel.sol

```diff
@@ -0,0 +1,117 @@
+    }
+
+    event MessageDispatched(uint64 nonce, bool result);
+
+    uint256 public constant MAX_GAS_PER_MESSAGE = 100000;
+    uint256 public constant GAS_BUFFER = 60000;
+
+    // Governance contracts will administer using this role.
+    bytes32 public constant CONFIG_UPDATE_ROLE =
+        keccak256("CONFIG_UPDATE_ROLE");
+
+    event RelayerNotRewarded(address relayer, uint256 amount);
+
+    RewardSource private rewardSource;
+
+    BeefyLightClient public beefyLightClient;
+
+    constructor(BeefyLightClient _beefyLightClient) {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        beefyLightClient = _beefyLightClient;
+        nonce = 0;
+    }
+
+    // Once-off post-construction call to set initial configuration.
+    function initialize(address _configUpdater, address _rewardSource)
+        external
+    {
+        require(
+            hasRole(DEFAULT_ADMIN_ROLE, msg.sender),
+            "Caller is unauthorized"
+        );
+
+        // Set initial configuration
+        grantRole(CONFIG_UPDATE_ROLE, _configUpdater);
+        rewardSource = RewardSource(_rewardSource);
+
+        // drop admin privileges
+        renounceRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function submit(
+        Message[] calldata _messages,
+        ParachainLightClient.ParachainVerifyInput
+            calldata _parachainVerifyInput,
+        ParachainLightClient.BeefyMMRLeafPartial calldata _beefyMMRLeafPartial,
+        uint256 _beefyMMRLeafIndex,
+        uint256 _beefyMMRLeafCount,
+        bytes32[] calldata _beefyMMRLeafProof
+    ) public {
+        // Proof
+        // 1. Compute our parachain's message `commitment` by ABI encoding and hashing the `_messages`
+        bytes32 commitment = keccak256(abi.encode(_messages));
+
+        ParachainLightClient.verifyCommitmentInParachain(
+            commitment,
+            _parachainVerifyInput,
+            _beefyMMRLeafPartial,
+            _beefyMMRLeafIndex,
+            _beefyMMRLeafCount,
+            _beefyMMRLeafProof,
+            beefyLightClient
+        );
+
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >= (_messages.length * MAX_GAS_PER_MESSAGE) + GAS_BUFFER,
+            "insufficient gas for delivery of all messages"
+        );
+
+        processMessages(payable(msg.sender), _messages);
+    }
+
+    function processMessages(
+        address payable _relayer,
+        Message[] calldata _messages
+    ) internal {
+        uint256 _rewardAmount = 0;
+
+        for (uint256 i = 0; i < _messages.length; i++) {
+            // Check message nonce is correct and increment nonce for replay protection
+            require(_messages[i].nonce == nonce + 1, "invalid nonce");
+
+            nonce = nonce + 1;
+
+            // Deliver the message to the target
+            // Delivery will have fixed maximum gas allowed for the target app
+            (bool success, ) = _messages[i].target.call{
```

Comment:
Could this lead to re-entrancy issues? 

Potentially problematic, since nonce is first incremented. 

Is call here still executed as internal? I assume not.

Reply:
same as above